apps/mark-scan/backend/src/custom-paper-handler/state_machine.ts

```diff
@@ -1,267 +1,394 @@
-    this.status = 'ejecting';
-    await this.driver.ejectBallot();
-  }
+/**
+ * Builds an observable that polls paper status and emits state machine events.
+ * Notes:
+ * Why Observable? This section from the xstate docs matches our use case closely:
+ * https://xstate.js.org/docs/guides/communication.html#invoking-observables
+ * "Observables can be invoked, which is expected to send events (strings or objects) to the parent machine,
+ * yet not receive events (uni-directional). An observable invocation is a function that takes context and
+ * event as arguments and returns an observable stream of events."
+ *
+ *
+ */
+function buildPaperStatusObservable() {
+  return ({ driver, pollingIntervalMs }: Context) => {
+    // `timer` returns an Observable that emits values with `pollingInterval` delay between each event
+    return (
+      timer(0, pollingIntervalMs)
+        // `pipe` forwards the value from the previous function to the next unary function ie. switchMap.
+        // In this case there is no value. The combination of timer(...).pipe() is so we can execute the
+        // function supplied to `switchMap` on the specified interval.
+        .pipe(
+          // `switchMap` returns an Observable that emits events.
+          // Why do we use `switchMap`? Is it because `switchMap` only emits items from the most recent
+          // inner Observable ie. it will only emit the latest PaperHandlerStatusEvent?
+          switchMap(async () => {
+            // Get raw status, map to event, and emit event
+            const paperHandlerStatus = await driver.getPaperHandlerStatus();
+            return paperHandlerStatusToEvent(paperHandlerStatus);
+          }),
+          timeout({
+            each: PAPER_HANDLER_STATUS_POLLING_TIMEOUT_MS,
+            with: () => throwError(() => new Error('paper_status_timed_out')),
+          })
+        )
+    );
+  };
+}
 
-  /**
-   * Ejects out the front. Can eject from loaded or parked state. If there is
-   * no paper to eject, handler will do nothing and return positive acknowledgement.
-   */
-  async ejectPaper(): Promise<void> {
-    this.status = 'ejecting';
-    await this.driver.ejectPaper();
-  }
+function pollPaperStatus(): InvokeConfig<Context, PaperHandlerStatusEvent> {
+  return {
+    id: 'pollPaperStatus',
+    src: buildPaperStatusObservable(),
+  };
+}
 
-  /**
-   * Moves paper to print position and moves print head to DOWN position. If
-   * paper is already in an appropriate print position, does not move paper.
-   * E.g. if paper is loaded, it will pull the paper in a few more inches, but
-   * if the paper is parked, it will not move the paper. Printing can start in
-   * a variety of positions.
-   */
-  async enablePrint(): Promise<void> {
-    await this.driver.enablePrint();
-  }
+const NoPaperState: StateNodeConfig<
+  Context,
+  any,
+  PaperHandlerStatusEvent,
+  BaseActionObject
+> = {
+  invoke: pollPaperStatus(),
+  on: {
+    PAPER_INSIDE: {
```

Comment:
How does it happen that the status goes straight from NO_PAPER to PAPER_INSIDE without going through PAPER_READY_TO_LOAD first?

Reply:
Good call, this is left over from before I separated out `LoadingPaperState` from `NoPaperState`. This event handler should have been deleted before requesting review, sorry!